edx_prefectutils/amplitude.py

```diff
@@ -0,0 +1,347 @@
+    if not old_high_watermark:
+        old_high_watermark = 0
+
+
+def _get_new_high_watermark(sf_credentials: SFCredentials, sf_role: str):
+    """
+    TODO: make this more robust (e.g. what if the table does not exist?  what if the table is empty?)
+    """
+    sf_connection = create_snowflake_connection(sf_credentials, sf_role)
+    cursor = sf_connection.cursor()
+    # Get the current maximum amplitude_event_id.
+    query = """
+        select max(amplitude_event_id)
+          from prod.amplitude_events.amplitude_events
+        """
+    cursor.execute(query)
+    return cursor.fetchone()[0]
+
+
+def get_processing_segments():
+    # Get the old (i.e. current) high watermark.
+    old_high_watermark = _get_old_high_watermark()
+
+    # Get the new high watermark, i.e. the current maximum amplitude_event_id.
+    new_high_watermark = _get_new_high_watermark(sf_credentials, sf_role)
+
+    # Get a list of processing segments which cover all events between the old and
+    # new high watermark. The result is a list of start/end tuples describing
+    # amplitude_event_id ranges, where each range represents a single processing
+    # segment with the configured batch size.  The start ID is inclusive, whereas
+    # the end ID is exclusive.
+    #
+    # e.g. if old=1000000, new=3600000, bucketsize=1000000, then:
+    # processing_segments = [
+    #     (1000000, 2000000),
+    #     (2000000, 3000000),
+    #     (3000000, 3600001),
+    # ]
+    #
+    # In the above example, the last batch is smaller than 1m events because there
+    # are no more events to include.
+    return [
+        (processing_segment_start, min(processing_segment_start + processing_bucket_size, new_high_watermark + 1))
+        for processing_segment_start in range(old_high_watermark, new_high_watermark, processing_bucket_size)
+    ]
+
+
+def load_events_into_amplitude(amplitude_api_key: str, processing_segments: list)
+    # Iterate over processing segments in chronological order, handling each one
+    # one at a time, and completely freeing the memory of the previous one before
+    # fetching the next.
+    for processing_segment in processing_segments:
+        # Prep a cursor to fetch the events for the current processing segment.
+        #
+        # For events with amplitude_event_id within the current processing_segment, further segment the events by the
+        # following dimension:
+        #
+        #   coalesce(user_id, device_id)
+        #
+        # Then, sort events by segment size and sort by timestamp within each segment.
+        segmented_by_device_user_query = (
+            f"""
+              select coalesce(user_id, device_id) as user_device,
+                     count(*) over (partition by user_device) as user_device_event_count,
+                     *
+                from events_to_load
+               where amplitude_event_id >= {processing_segment_start}
+                 and amplitude_event_id < {processing_segment_end}
```

Comment:
So if we want this to match other places we use watermarks, I think this would be 
```where amplitude_event_id >= {processing_segment_start}  and amplitude_event_id <= {processing_segment_end}```

Reply:
I've retained pythonic behavior with processing segment boundaries, but adopted organizationally-consistent behavior with the watermarks.